app/services/action_builder.rb

```diff
@@ -20,6 +20,15 @@ def build_action(extra_attrs = {})
   def previous_action
```

Comment:
For campaign pages, i'd actually collect/pluck the page ids first, then do the lookup for the action.